File: src/core/options.rs
use std::env::args;

#[derive(PartialEq, Copy, Clone)]
pub enum OptionKind {
    Bad,

    Init, Build, Help, Version,

    SourceFile, OutputFile,
    
    ShowNodes, NoOpt,
}

pub struct ApicOption {
    kind: OptionKind,
    value: String,
}

pub fn contain_option_kind(options: &Vec<ApicOption>, option_kind: OptionKind) -> bool {
    for option in options {
        if *option.get_kind() == option_kind {
            return true;
        }
    }
    
    false
}

impl ApicOption {
    pub fn init(kind: OptionKind, value: String) -> ApicOption {
        ApicOption { kind, value }
    }

    pub fn get_kind(&self) -> &OptionKind {
        &self.kind
    }

    pub fn get_value(&self) -> &String {
        &self.value
    }
}

pub fn handle_options() -> Vec<ApicOption> {
    let args: Vec<String> = args().collect();
    let mut options = vec![];

    if args.len() < 2 {
        options.push(ApicOption::init(OptionKind::Bad, String::from("apic error : use `apic help` to see how to use the apica compiler")));
        return options;
    }

    if !handle_command(&mut options, &args) {
        return options;
    }

    if args.len() > 2 {
        let mut awaiting: Option<OptionKind> = None;
        for arg in &args[2..] {
            if arg.chars().nth(0).unwrap() == '-' {
                if awaiting.is_some() {
                    options.insert(0, ApicOption::init(OptionKind::Bad, String::from("apic error: cannot use an option right after an option that awaited a value. Use `apic help` to see how to use the apica compiler")));
                    return options;
                }

                match arg.as_str() {
                    "-i" => awaiting = Some(OptionKind::SourceFile),
                    "-o" => awaiting = Some(OptionKind::OutputFile),
                    "--nodes" => options.push(ApicOption::init(OptionKind::ShowNodes, String::new())),
                    "--no-opt" => options.push(ApicOption::init(OptionKind::NoOpt, String::new())),

                    _ => {
                        options.insert(0, ApicOption::init(OptionKind::Bad, String::from("apic error: invalid option found. Use `apic help` to see how to use the apica compiler")));
                        return options;
                    }
                }
            } else {
                if let Some(kind) = &awaiting {
                    options.push(ApicOption::init(*kind, arg.clone()));
                    awaiting = None;
                } else {
                    options.insert(0, ApicOption::init(OptionKind::Bad, String::from("apic error: a value can only be placed after an option that requires one. Use `apic help` to see how to use the apica compiler`")));
                    return options;
                }
            }
        }
    }

    options
}

fn handle_command(options: &mut Vec<ApicOption>, args: &Vec<String>) -> bool {
    match args[1].as_str() {
        "help" => options.push(ApicOption::init(OptionKind::Help, String::new())),
        "init" => options.push(ApicOption::init(OptionKind::Init, String::new())),
        "build" => options.push(ApicOption::init(OptionKind::Build, String::new())),
        "version" => options.push(ApicOption::init(OptionKind::Version, String::new())),

        _ => {
            options.push(ApicOption::init(OptionKind::Bad, String::from("apic error : incorrect command found. Use `apic help` to see the availabled commands")));
            return false;
        }
    }

    true
}

File: src/core/options.cpp
#include "core/options.hpp"
#include "utils/errors.hpp"
#include <cstring>

using namespace core;

struct AwaitedOptionKind {
    std::optional<ApicOptionKind> kind;
    std::string cmd;
};

ApicOptions::ApicOptions() {

}

ApicOptions ApicOptions::getOptions(int argc, char **argv) {
    ApicOptions apic_options;
    if (argc < 2) {
        apic_options.addCommandOrError(ApicOptionKind::Bad, std::string(utils::OPT_ERROR_NO_ARGUMENT));
        return apic_options;
    }

    if (!apic_options.handleCommand(argv[1])) {
        apic_options.addCommandOrError(ApicOptionKind::Bad, std::string(utils::OPT_ERROR_UNKNOWN_COMMAND));
        return apic_options;
    }

    AwaitedOptionKind awaited_kind;
    awaited_kind.cmd = "";
    awaited_kind.kind = std::nullopt;

    for (int i = 2; i < argc; i++) {
        if (argv[i][0] == '-') {
            if (awaited_kind.kind) {
                std::string error_message(utils::OPT_ERROR_OPTION_EXPECTED_VALUE);
                error_message += awaited_kind.cmd;
                error_message += '`';

                apic_options.addCommandOrError(ApicOptionKind::Bad, error_message);
                return apic_options;
            }

            if (!strcmp(argv[i], "-i")) {
                awaited_kind.kind = ApicOptionKind::SourceFile;
                awaited_kind.cmd = "-i";
            } else if (!strcmp(argv[i], "-o")) {
                awaited_kind.kind = ApicOptionKind::OutputFile;
                awaited_kind.cmd = "-o";
            } else if (!strcmp(argv[i], "--nodes")) {
                if (!apic_options.addOption(ApicOptionKind::ShowNodes, "", "--nodes"))
                    return apic_options;
            } else if (!strcmp(argv[i], "--werror")) {
                if (!apic_options.addOption(ApicOptionKind::WarningsAsErrors, "", "--werror"))
                    return apic_options;
            } else {
                std::string error_message(utils::OPT_ERROR_INVALID_OPTION);
                error_message += argv[i];
                error_message += '`';

                apic_options.addCommandOrError(ApicOptionKind::Bad, error_message);
                return apic_options;
            }
        } else {
            if (awaited_kind.kind) {
                if (!apic_options.addOption(awaited_kind.kind.value(), argv[i], awaited_kind.cmd))
                    return apic_options;
                awaited_kind.kind = std::nullopt;
            } else {
                std::string error_message(utils::OPT_ERROR_VALUE_UNEXPECTED);
                error_message += argv[i];
                error_message += '`';

                apic_options.addCommandOrError(ApicOptionKind::Bad, error_message);
                return apic_options;
            }
        }
    }

    if (awaited_kind.kind) {
        std::string error_message(utils::OPT_ERROR_AWAITED_VALUE);
        error_message += awaited_kind.cmd;
        error_message += '`';

        apic_options.addCommandOrError(ApicOptionKind::Bad, error_message);
    }

    return apic_options;
}

std::optional<std::string> ApicOptions::getOption(ApicOptionKind kind) const {
    auto option = this->options.find(kind);
    if (option == this->options.end())
        return std::nullopt;
    
    return option->second;
}

bool ApicOptions::handleCommand(char *arg) {
    if (!strcmp(arg, "help")) this->addCommandOrError(ApicOptionKind::Help, "");
    else if (!strcmp(arg, "init")) this->addCommandOrError(ApicOptionKind::Init, "");
    else if (!strcmp(arg, "build")) this->addCommandOrError(ApicOptionKind::Build, "");
    else if (!strcmp(arg, "version")) this->addCommandOrError(ApicOptionKind::Version, "");
    else if (!strcmp(arg, "lsp")) this->addCommandOrError(ApicOptionKind::Lsp, "");
    else return false;

    return true;
}

void ApicOptions::addCommandOrError(ApicOptionKind kind, const std::string &value) {
    this->options[kind] = value;
}

bool ApicOptions::addOption(ApicOptionKind kind, const std::string &value, const std::string &cmd) {
    auto option = this->options.find(kind);
    if (option != this->options.end()) {
        std::string error_message(utils::OPT_ERROR_ALREADY_DEFINED);
        error_message += cmd;
        error_message += '`';

        this->addCommandOrError(ApicOptionKind::Bad, error_message);
        return false;
    }

    this->options[kind] = value;
    return true;
}